src/Interpreters/DatabaseCatalog.h

```diff
@@ -203,6 +205,12 @@ class DatabaseCatalog : boost::noncopyable, WithMutableContext
     /// this method will throw an exception.
     void addUUIDMapping(const UUID & uuid);
 
+    bool hasUUIDMapping(const UUID & uuid);
+
+    void addProtectedUUIDDir(const UUID & uuid);
+    void removeProtectedUUIDDir(const UUID & uuid);
```

Comment:
Actually it was ugly ad-hoc trick, I've removed `addProtectedUUIDDir` and `removeProtectedUUIDDir` methods and added more assertions (so it looks less bad now).

As for `noexcept`, it's really helpful sometimes (for example, when you would prefer to get `SIGABRT` instead of inconsistent state of data structures in memory due to some stupid exception like `bad_alloc`), but in most cases it does not make sense. I don't think `noexcept` is needed for method that calls only `std::unordered_set::erase`. Also it was called only from dtor that is `noexcept` be default.